### ISSUE TITLE
Smart expander I:  Breadth-first traversal of the parsed token sequence

### ISSUE THREAD

**Issue by clarity20** [clarity20/tql#8]
With the parser nearing completion, our attention turns more and more to the smart expansion phase. We have a prototype "dumb expander" in place for the SELECT clause, and a WHERE-clause version would be similar. The smart expander would take the completion suggestions offered by its dumb predecessor and filter out those candidates whose "data types" are incompatible with the grammar of the containing expression. So, for instance, we would discard a string-type completion candidate where only a floating-point candidate would make grammatical sense.

Our first step toward this goal will be to correctly traverse the parse "tree" so that we can reason correctly about the global grammatical implications of local constructs within the expression. This means a breadth-first traversal where ties are broken in such a way that order of operations is respected, multiplications before additions _et cetera_.

In other words, we will break down the lexical entities into operators and operands at every relevant level of detail, from the topmost to the atomic (_i.e._ single tokens). The implementation will be recursive. Testing will involve outputting the operators with their operands as we visit them. The order of the outputs and the content of each one both need to be correct.

**Comment by clarity20** [comment]
A more basic precursor to the above is the following:

Adjust the parsing routines to pass along the depth (_i.e._ the global parenthesis nesting level) of each operator as part of the operator-token's markup.

Do this first before tackling the above.

**Comment by clarity20** [comment]
Checkin # 7066a6f adds the nesting depth to the operator and comparator markup.

**Comment by clarity20** [comment]
Commit # 4a7db9e includes a function that traverses the parse tree (in postorder fashion) to determine the grammatical legality of a set of completion choices. Some simplifying assumptions were made in constructing this function; this was the intention all along. Right now:

- Only column names can be completed. 
- Parenthesized subexpressions cannot be processed. 
- The lists of grammar rules are not exhaustive. 
- Abbreviated words are not allowed to be literal. 

But all of this is intentional. For this Issue we intend only to shoot the first, long strands of catenary wire across the river from one concrete anchor to the other, draped over the supporting towers in between. Future issues will address laying the roadbed and opening it to traffic.

**Comment by clarity20** [comment]
Commit #207c1f2 ties the loose ends together: with the feasibility of each choice combination having been assessed and recorded, the lists of _feasible_ choices at each coordinate are presented to the user one-by-one. Each coordinate choice triggers further pruning (if possible), and the abbreviated expression is inflated to its full-bodied equivalent.

**Comment by clarity20** [comment]
Commit # 6600257 contains fixes to this functionality facilitated by the new test data added in # fa87f64.
